caffe2/core/storage.h

```diff
@@ -77,117 +80,135 @@ class CAFFE2_API StorageImpl : public c10::intrusive_ptr_target {
+    std::swap(data_ptr_, data_ptr);
+    return std::move(data_ptr);
+  };
+
+  void set_dtype(const TypeMeta& data_type) {
+    int64_t capacity = numel_ * data_type_.itemsize();
     data_type_ = data_type;
+    numel_ = capacity / data_type_.itemsize();
   }
 
-  const DataType& dtype() const {
+  const TypeMeta& dtype() const {
     return data_type_;
   }
 
+  const at::Allocator* allocator() const {
+    return allocator_;
+  };
+  // You generally shouldn't use this method, but it is occasionally
+  // useful if you want to override how a tensor will be reallocated,
+  // after it was already allocated (and its initial allocator was
+  // set)
+  void set_allocator(at::Allocator* allocator) {
+    allocator_ = allocator;
+  }
+
   size_t capacity() const {
-    return capacity_;
+    return numel_ * itemsize();
   }
 
   int64_t numel() const {
-    return capacity_ / itemsize();
+    return numel_;
   }
 
   // TODO: remove later
   void set_numel(int64_t numel) {
-    capacity_ = numel * itemsize();
+    numel_ = numel;
   }
 
-  inline DeviceType device_type() const {
-    return device_type_;
+  at::DeviceType device_type() const {
+    return data_ptr_.device().type();
   }
 
   inline size_t itemsize() const {
     return data_type_.itemsize();
   }
 
-  // Rule of Five
-  StorageImpl(StorageImpl&&) = default;
-  ~StorageImpl() = default;
-  StorageImpl& operator=(StorageImpl&&) = default;
+  bool resizable() const {
+    return resizable_;
+  };
+
+  void set_resizable(bool resizable) {
+    resizable_ = resizable;
+  }
 
   /**
    * Can only be called when use_count is 1
    */
-  template <typename Deleter = MemoryDeleter>
   void UniqueStorageShareExternalPointer(
       void* src,
-      const DataType& data_type,
+      const TypeMeta& data_type,
       size_t capacity,
-      Deleter d = nullptr) {
+      MemoryDeleter d = nullptr) {
+    UniqueStorageShareExternalPointer(
+        at::DataPtr(src, src, d, data_ptr_.device()), data_type, capacity);
+  }
+
+  /**
+   * Can only be called when use_count is 1
+   */
+  void UniqueStorageShareExternalPointer(
+      at::DataPtr&& data_ptr,
+      const TypeMeta& data_type,
+      size_t capacity) {
     data_type_ = data_type;
     CAFFE_ENFORCE_WITH_CALLER(
         data_type_.id() != TypeIdentifier::uninitialized(),
         "To share with a raw external pointer you need to have meta "
         "already set.");
-    // Check if the deleter is a MemoryDeleter and is a simple nullptr.
-    if (std::is_same<MemoryDeleter, Deleter>::value &&
-        reinterpret_cast<MemoryDeleter*>(&d)[0] == nullptr) {
-      // Use aliasing constructor trick to avoid calling the destructor.
-      data_ptr_ = std::shared_ptr<void>(std::shared_ptr<void>(), src);
-    } else {
-      data_ptr_.reset(src, d);
-    }
-    capacity_ = capacity;
+    data_ptr_ = std::move(data_ptr);
+    // NOTE: data_type might change and so it's also possible that capacity
```

Comment:
I think this comment is out of date. It is true that `data_type` may change, but in this case, we're simultaneously changing `data_type` while we're swapping out the data pointer, so we don't care about the old capacity; we're just setting the capacity based on what the user told us is the new capacity (in bytes!!) It might not be divisible by itemsize(), but then that's because the user gave us a weirdly rounded region of memory.